src/indexing.jl

```diff
@@ -488,6 +483,80 @@ end
     end
 end
 
+
+_getindex_kwargs(x, kwargs, args...) = @inbounds getindex(x, args...; kwargs...)
+
+function _generate_unsafe_getindex!_body(N::Int)
+    quote
+        Base.@_inline_meta
+        D = eachindex(dest)
+        Dy = iterate(D)
+        @inbounds Base.Cartesian.@nloops $N j d->I[d] begin
+            # This condition is never hit, but at the moment
+            # the optimizer is not clever enough to split the union without it
+            Dy === nothing && return dest
+            (idx, state) = Dy
+            dest[idx] = Base.Cartesian.@ncall $N _getindex_kwargs src kwargs j
+            Dy = iterate(D, state)
+        end
+        return dest
+    end
+end
+
+@generated function _unsafe_getindex!(dest::AbstractArray, src::AbstractArray, I::Vararg{Union{Real, AbstractArray}, N}; kwargs...) where N
+    _generate_unsafe_getindex!_body(N)
+end
+
+function unsafe_get_collection(::CPUIndex, src, inds; kwargs...)
+    return unsafe_get_collection_by_index(IndexStyle(src), src, inds)
+end
+
+# TODO
+function unsafe_get_collection_by_index(::IndexLinear, src, inds; kwargs...)
+    dst = similar(src, Base.index_shape(inds...))
+    src_iter = @inbounds(view(LinearIndices(src), inds...)) # FIXME not ideal iterator
```

Comment:
I assume that the default Base._generate_unsafe_getindex!_body isn't as efficient as it could be because it uses cartesian like APL indexing. I figure we could take advantage of arrays where we know the indexing style is linear. However, I'm not aware of an iterator that is ideal for this right now.

Reply:
It's probably pretty good.
Normally, loads are a lot more important for performance than stores. That's because lots of subsequent operations tend to be dependent on loads, but that wouldn't be the case for a store unless it's immediately reloaded. Meaning an out of order CPU normally won't have to wait on a store before executing future instructions.

But here, basically all the CPU is doing is loading and storing, so there isn't any other work the CPU can do instead.
And most CPUs can perform twice as many loads per cycle as stores.
Meaning the stores are probably more important here.

And, presumably `dest` is getting allocated to column major, so that this order is pretty good for it. And if we have basically anything other than `UnitRange`s in `I::Vararg{Union{Real, AbstractArray}, N}`, the access pattern on `src` is going to require slow `gather` instructions anyway.

Reply:
So the base implementation is probably as good as we can do for CPUIndex?

Reply:
Yeah, it's definitely most of the way there.